pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -205,21 +362,102 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
 		// requests may fail if we serve them in mixed order.
 		// TODO: may revisit this part later if we see inefficient resource allocation
 		// in real use as the result of this.
-		resp, err := h.devicePluginManager.Allocate(resource, append(devices.UnsortedList(), allocated...))
+		resp, err := h.devicePluginManager.Allocate(resource, allocDevices)
 		if err != nil {
-			return nil, err
+			return fmt.Errorf("requested number of devices unavailable for %s. Requested: %d", resource, needed)
 		}
-		ret = append(ret, resp)
+		// Update internal cached podDevices state.
+		h.Lock()
+		allocDeviceSet := sets.NewString(allocDevices...)
+		if oldDevices := h.podDevices.containerDevices(string(pod.UID), container.Name, resource); oldDevices != nil {
+			freeDevices := oldDevices.Difference(allocDeviceSet)
+			h.allocatedDevices[resource] = h.allocatedDevices[resource].Difference(freeDevices)
+		}
+		h.podDevices.insert(string(pod.UID), container.Name, resource, allocDeviceSet, *resp)
+		h.Unlock()
 	}
+
 	// Checkpoints device to container allocation information.
 	if err := h.writeCheckpoint(); err != nil {
-		return nil, err
+		return err
 	}
-	return ret, nil
+	return nil
 }
 
-// updateAllocatedDevices updates the list of GPUs in use.
-// It gets a list of active pods and then frees any GPUs that are bound to
+func (h *DevicePluginHandlerImpl) Allocate(node *schedulercache.NodeInfo, attrs *lifecycle.PodAdmitAttributes) error {
+	pod := attrs.Pod
+	pods := attrs.OtherPods
+	// TODO: determine whether to take initContainer into consideration
+	for _, container := range pod.Spec.Containers {
+		if devOpts := h.GetDeviceRunContainerOptions(pod, &container); devOpts != nil {
+			continue
+		}
+		// No runtime information found in cache, make allocation
+		if err := h.allocateContainerResources(pod, &container); err != nil {
+			return err
+		}
+
```

Comment:
Suggest we add a variable podRequireDevicePluginResource := false at the beginning of the function and update it here with:
_, ok := h.podDevices[pod.podUID]
podRequireDevicePluginResource |= ok

This variable can be used later to avoid performance overhead of sanitizeNodeAllocatableResource() on the common case that pod doesn't require any device plugin resource.